File: SimGenerator.py
import sim_modules as sm
import numpy as np
import matplotlib.pyplot  as plt
import sys 
import input as input 
import os
size= input.size
pathin= input.pathin
pathout=input.pathout
cols = 2*size
print('Sim Directory is %s'%input.pathin)
print('reading from ', input.num_participants, 'participants', input.steps, 'timesteps', input.ampContemp, input.ampLagged, input.ampMeasure, input.maskZero)
printpath=pathout+'sims/numParticipants_%s'%input.num_participants
print('Main dir is %s'%printpath)

plt.close()

savepath=pathout+'sims/numParticipants_%s/steps_%s/Contemp_%s_amp_%s/Lagged_%s_amp_%s/Measure_%s_amp_%s/mask_%s/clipsigma_%s/'%(input.num_participants,str(input.steps),input.covContempName,str(input.ampContemp),input.covLaggedName, str(input.ampLagged),input.measurecovName, str(input.ampMeasure),str(input.maskZero),str(input.clip_outliers))
try:
    os.makedirs(savepath)
    if input.debug:
        print('making %s'%savepath)
except:
    if input.debug:
        print('directory %s exists'%savepath)
    pass
if input.clip_outliers:
    if input.debug:
        print('Making clip/preclip dirs')
    savepathprecliporig=os.path.join(savepath,'preclip/')
    try:
        os.mkdir(savepathprecliporig)
    except:
        if input.debug:
            print('preclip dir exists')
    
    savepathcliporig=os.path.join(savepath,'clip/')
    try:
        os.mkdir(savepathcliporig)
    except:
        if input.debug:
            print('clip dir exists')

for j in range(input.num_iterations):
            
    if input.debug:
        print('making rep dirs')
    savepathclip=os.path.join(savepathcliporig,'rep_%i/'%j)
    savepathpreclip=os.path.join(savepathprecliporig,'rep_%i/'%j)
    try:
        os.mkdir(savepathclip)
        os.mkdir(savepathpreclip)
    except:
        print(f'rep dir {j} exists')
    #vec = [97, 68, 6, 40, 41, 7, 69, 96, 94, 43, 5, 4, 42, 95, 91, 1, 92, 3, 45, 44, 2, 93, 122, 37, 36, 123, 121, 135, 34, 35, 134, 124, 130, 31, 131, 125, 133, 127, 32, 33, 126, 132, 103, 102, 100, 128, 15, 14, 129, 101, 105, 10, 38, 39, 11, 104, 13, 12, 75, 61, 74, 62, 63, 98, 67, 73, 9, 8, 72, 66, 99, 70, 64, 65, 71]
    for number in range(input.num_participants):
    #for number in vec:
        
        csvnum = number+1
     #   csvnum = number
        if input.debug:
            print('Writing sims for participant %i and sim iteration %i'%(csvnum,j))
        file = pathin+'%i.csv'%csvnum
        data = np.loadtxt(file, skiprows=1, usecols=range(0,cols), delimiter=',')
        matContemp = data[:,size:] #same day (contemporaneous) beta values
        matLagged = data[:,:size] #lagged beta values
        
        # sm.save_matrix(matContemp, 'matContempOrig.csv')
        # sm.save_matrix(matLagged, 'matLaggedOrig.csv')
        # sm.plot_matrix(matLagged, True, 'matLaggedOrig.png')
        # sm.plot_matrix(matContemp, True, 'matContempOrig.png')
        
        if input.covContempName=='randn':
            covContemp = np.random.randn(size,size)
        else:
            covContemp = np.ones(size)

        if input.covLaggedName=='randn':
            covLagged = np.random.randn(size,size)
        else:
            covLagged =np.ones(size)

        if input.measurecovName=='diag':
            measureCov= input.ampMeasure*np.eye(size)
        
        if input.maskZero:
            maskContemp =  sm.make_mask(matContemp, contemp=True)
            maskLagged = sm.make_mask(matLagged, contemp=False)
        else:
            maskContemp =  np.ones((input.size, input.size))
            np.fill_diagonal(maskContemp,0) # still make diagonal zeros
            maskLagged = np.ones((input.size, input.size))

        # sm.plot_matrix(maskLagged, True, 'maskLagged.png')
        # sm.plot_matrix(maskContemp, True, 'maskContemp.png')
        plt.close()
        samples = sm.generate_timeseries(input.start, input.steps, input.ampContemp, matContemp, covContemp, input.ampLagged, matLagged, covLagged,measureCov,input.save)
        
        
        if input.debug:
            print(np.sqrt(np.diagonal(measureCov)[0])*3, 'max')
            for p in range(np.shape(samples)[1]):

                inds_pre = np.where(samples[:,p]>np.sqrt(np.diagonal(measureCov)[0])*3)[0]
                print(inds_pre, samples[inds_pre,p],p)
            print('---- TEST pre clip -----')
        if input.clip_samples:
        # Checking the clipping
            samples_clip = sm.clip_timeseries(samples, input.clip_indices, input.clip_mins, input.clip_maxs)
            samples=samples_clip
            if input.debug:
                print('saved clip')
        
        if input.clip_outliers:
            if input.debug:
                print('Clipping samples if needed')
            
            savefile = savepathpreclip+'ind_%i_preclip.txt'%(csvnum)
            np.savetxt(savefile,samples, delimiter=',')
            samples_clip = sm.clip_outliers(samples, input.clip_sigma,input.ampMeasure, input.debug,input.start, input.ampContemp, matContemp, covContemp, input.ampLagged, matLagged, covLagged,measureCov)
            samples=samples_clip.copy()
            if input.debug:
                for p in range(np.shape(samples)[1]):
                    inds_post = np.where(samples[:,p]>np.sqrt(np.diagonal(measureCov)[0])*3)[0]
                    print(inds_post,samples[inds_post,p],p)

                print('---- TEST post clip -----')
            if input.debug:
                print('saved clip')
            savefile = savepathclip+'ind_%i.txt'%(csvnum)
        
        np.savetxt(savefile,samples, delimiter=',')

File: sim_generator.R
# Script to generate and store simulated pre-clipped and clipped timeseries data for all individuals 

# Set working directory
setwd("/Users/reneehlozek/Code/Collaborative-CIFAR-Project-/")
library('SyncRNG')
seed=123456
s <- SyncRNG(seed=seed)
universal_rng <- SyncRNG(seed=seed)
# Load scripts
source("input.R")
source("sim_modules.R")


# Initialize input variables
size <- size
pathin <- pathin
pathout <- pathout
cols <- 2 * size
steps <- steps 
print(steps)
num_participants <- num_participants
num_iterations <- num_iterations
save <- save
covContempName <- covContempName
covLaggedName <- covLaggedName
measurecovName <- measurecovName
ampContemp <- ampContemp
ampLagged <- ampLagged
ampMeasure <- ampMeasure
maskZero <- maskZero
clipSamples <- clipSamples
clipIndices <- clipIndices
clipMins <- clipMins
clipMaxs <- clipMaxs
clipOutliers <- clipOutliers
clipSigma <- clipSigma

print(paste("Sim Directory is", pathin))
print(paste("reading from", num_participants, "participants", steps, "timesteps", ampContemp, ampLagged, ampMeasure, maskZero))
printpath <- file.path(pathout, "sims", paste0("numParticipants_", num_participants))
print(paste("Main dir is", printpath))

#dev.off()

# Make directory for sims
# Example: /Users/mac/Desktop/CIFAR/220429.FinalMatrices_sims/sims/numParticipants_15/steps_10/Contemp_randn_amp_0.1/Lagged_randn_amp_0.1/Measure_diag_amp_1/Mask_TRUE/clipsigma_3
savepath <- file.path(pathout, "sims")
savepath <- file.path(savepath, paste0("numParticipants_", num_participants))
savepath <- file.path(savepath, paste0("steps_", steps))
savepath <- file.path(savepath, paste0("Contemp_", covContempName, "_amp_", ampContemp))
savepath <- file.path(savepath, paste0("Lagged_", covLaggedName, "_amp_", ampLagged))
savepath <- file.path(savepath, paste0("Measure_", measurecovName, "_amp_", ampMeasure))
savepath <- file.path(savepath, paste0("Mask_", maskZero))
savepath <- file.path(savepath, paste0("clipsigma_", clipSigma))
cat("Save path is", savepath, "\n")


# Make directories if they do not exist
if (!dir.exists(savepath)) {
  dir.create(savepath, recursive = TRUE)
  # if (debug) { cat("making", savepath, "\n") }
  }

# Check and create clip and preclip directories if clip_outliers is TRUE
if (clipOutliers) {
  # if (debug) {cat("Making clip/preclip dirs\n")}
  savepathprecliporig <- file.path(savepath, "preclip/")
  savepathcliporig <- file.path(savepath, "clip/")
  dir.create(savepathprecliporig, showWarnings = FALSE, recursive = TRUE)
  dir.create(savepathcliporig, showWarnings = FALSE, recursive = TRUE)
}

# Loop over iterations
for (j in 1:num_iterations) {
  # Set unique seed for each iteration to control randomness 
  seed = seed+j
  s <- SyncRNG(seed=seed)
  universal_rng <- SyncRNG(seed=seed)
  
  # if (debug) {
  #   cat("making rep dirs\n")
  # }
  
  savepathclip <- file.path(savepathcliporig, paste0("rep_", j, "/"))
  savepathpreclip <- file.path(savepathprecliporig, paste0("rep_", j, "/"))
  dir.create(savepathclip, showWarnings = FALSE, recursive = TRUE)
  dir.create(savepathpreclip, showWarnings = FALSE, recursive = TRUE)
  
  # Loop over participants
  for (number in 1:num_participants) {
    # if (debug) {
    #   cat("Writing sims for participant", number, "and sim iteration", j, "\n")
    # }
    
    # printpath(pathin)
    # print(huh)
    # Load in input contemp and lagged coeff matrix data
    file <- file.path(pathin, paste0("ind_",number, ".csv"))

    data <- read.csv(file, header = TRUE, sep = ",")
    matContemp <- as.matrix(data[, (size + 1):cols])
    matLagged <- as.matrix(data[, 1:size])
    
    # Generate random covariance matrices
    if (covContempName == "randn") {
      covContemp <- matrix(rnorm(size * size), nrow = size, ncol = size)
    } else {
      covContemp <- matrix(1, nrow = size, ncol = size)
    }
    
    if (covLaggedName == 'randn') {
      covLagged <- matrix(rnorm(size * size), nrow = size, ncol = size)
    } else {
      covLagged <- matrix(1, nrow = size, ncol = size)
    }
    
    # Create measureCov matrix 
    if (measurecovName == 'diag') {
      measureCov <- ampMeasure * diag(size)
    } else {
      measureCov <- matrix(0, nrow = size)
    }
    
    # Apply mask
    maskContemp <- ifelse(maskZero, make_mask(matContemp, contemp = TRUE), matrix(1, nrow = size, ncol = size))
    maskLagged <- ifelse(maskZero, make_mask(matLagged, contemp = FALSE), matrix(1, nrow = size, ncol = size))
    
    # Generate timeseries data
    samples <- generate_timeseries(start, steps, ampContemp, matContemp, covContemp,ampLagged, matLagged, covLagged, measureCov, save)
    # print(nrow(samples))
    # print(ncol(samples))
    
    max_clip_threshold <- sqrt(diag(measureCov)[1]) * 3
    
    # if (debug) {
    #   cat(max_clip_threshold, 'max', "\n")
    #   for (p in 1:ncol(samples)) {
    #     inds_pre <- which(samples[, p] > max_clip_threshold)
    #     cat("pre clip:", inds_pre, samples[inds_pre, p], p, "\n")
    #   }
    #   cat("---- TEST pre clip -----\n")
    # }
    
    # Clip time series data if specified
    if (clipSamples) {
      # Checking the clipping
      samples_clip <- clip_timeseries(samples, clip_indices, clip_mins, clip_maxs)
      samples <- samples_clip
      # if (debug) {
      #   cat('saved clip\n')
      # }
    }
    
    if (clipOutliers) {
      # if (debug) {
      #   cat('Clipping samples if needed\n')
      # }
      
      savefile <- file.path(savepathpreclip, paste0('ind_', number, '_preclip.txt'))
      write.table(samples, file = savefile, sep = ',', row.names = FALSE, col.names = FALSE)
      
      samples_clip <- clip_outliers(
        samples, clip_sigma, ampMeasure, debug, start,
        ampContemp, matContemp, covContemp, ampLagged, matLagged,
        covLagged, measureCov
      )
      
      samples <- samples_clip
      
      #  if (debug) {
      #   for (p in 1:ncol(samples)) {
      #     inds_post <- which(samples[, p] > max_clip_threshold)
      #     cat("post clip:", inds_post, samples[inds_post, p], p, "\n")
      #   }
      #   cat("---- TEST post clip -----\n")
      #   cat('saved clip\n')
      # }
    
      savefile <- file.path(savepathclip, paste0('ind_', number, '.txt'))
    }
    
    write.table(samples, file = savefile, sep = ',', row.names = FALSE, col.names = FALSE)
  }
}



